File: 2022/day-3.hs
import Data.List.Split (splitOn)
import Data.Set (Set)
import qualified Data.Set as Set
import Data.Ix (inRange)
import Data.Char (ord)

filterEmptyString :: [String] -> [String]
filterEmptyString = filter ((> 0) . length)

parseInput :: String -> [[Char]]
parseInput = filterEmptyString . splitOn "\n"

getCompartments :: String -> (String, String)
getCompartments s = (take (div (length s) 2) s, drop (div (length s) 2) s)

getShared :: (String, String) -> [Char]
getShared (s1, s2) = Set.toList $ Set.intersection (Set.fromList s1) (Set.fromList s2)

getPriority :: Char -> Int
getPriority c | (inRange ('a','z') c) = (ord c) - (ord 'a') + 1
              | (inRange ('A','Z') c) = (ord c) - (ord 'A') + 27

chunk :: Integer -> [a] -> [[a]]
chunk n [] = []
chunk n xs = (take 3 xs):(chunk n $ drop 3 xs)

main = do
  contents <- readFile $ "day-3-input.txt"
  let input = parseInput contents

  let result1 = sum $ input >>= ((map getPriority) . getShared . getCompartments)
  print $ "Part 1: " ++ show (result1)

  let result2 = sum $ chunk 3 input >>= (map getPriority) . (foldl1 (\a b -> getShared (a,b)))
  print $ "Part 2: " ++ show (result2)

File: 2022/day-3.ex
input = File.read!("day-3-input.txt")

defmodule AoC do
  def get_compartments(s) do
    String.split_at(s, div(String.length(s), 2))
  end

  def get_shared({s1, s2}) do
    MapSet.intersection(
      MapSet.new(String.to_charlist(to_string(s1))),
      MapSet.new(String.to_charlist(to_string(s2)))
    )
    |> MapSet.to_list()
  end

  def get_priority(c) do
    cond do
      c in ?a..?z -> c - ?a + 1
      c in ?A..?Z -> c - ?A + 27
    end
  end
end

data =
  input
  |> String.trim()
  |> String.split("\n")

result1 =
  data
  |> Enum.map(&AoC.get_compartments/1)
  |> Enum.flat_map(&AoC.get_shared/1)
  |> Enum.map(&AoC.get_priority/1)
  |> Enum.sum()

IO.puts("Part1: #{result1}")

result2 =
  data
  |> Enum.chunk_every(3)
  |> Enum.flat_map(fn chunk -> Enum.reduce(chunk, fn a, b -> AoC.get_shared({a, b}) end) end)
  |> Enum.map(&AoC.get_priority/1)
  |> Enum.sum()

IO.puts("Part2: #{result2}")
